torchdata/datapipes/iter/util/combining.py

```diff
@@ -213,3 +216,139 @@ def __iter__(self) -> Iterator:
 
     def __len__(self) -> int:
         return len(self.source_iterdatapipe)
+
+
+def _drop_index(idx_data):
+    idx, data = idx_data
+    return data
+
+
+@functional_datapipe("round_robin_demux")
+class RoundRobinDemultiplexerIterDataPipe(IterDataPipe):
+    r"""
+    Splits the input DataPipe into multiple child DataPipes in the round-robin order (functional name: ``round_robin_demux``).
+    A list of the child DataPipes is returned from this operation.
+
+    Args:
+        datapipe: Iterable DataPipe being filtered
+        num_instances: number of instances of the DataPipe to create
+        buffer_size: this defines the maximum number of inputs that the buffer can hold across all child
+            DataPipes while waiting for their values to be yielded.
+            Defaults to ``1000``. Use ``-1`` for the unlimited buffer.
+
+    Examples:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> def odd_or_even(n):
+        ...     return n % 2
+        >>> source_dp = IterableWrapper(range(5))
+        >>> dp1, dp2 = source_dp.round_robin_demux(2)
+        >>> list(dp1)
+        [0, 2, 4]
+        >>> list(dp2)
+        [1, 3]
+    """
+
+    def __new__(cls, datapipe: IterDataPipe, num_instances: int, buffer_size: int = 1000):
+        if num_instances < 1:
+            raise ValueError(f"Expected `num_instaces` larger than 0, but {num_instances} is found")
+        if num_instances == 1:
+            warnings.warn(
+                "The operation of `round_robin_demux` with `num_instances=1` is an no-op and returns the provided `datapipe` directly"
+            )
+            return datapipe
+
+        datapipe = datapipe.enumerate()
+        container = _RoundRobinDemultiplexerIterDataPipe(datapipe, num_instances, buffer_size=buffer_size)
```

Comment:
curious: can we just do 
```python
_DemultiplexerIterDataPipe(
    datapipe, 
    num_instances, 
    partial(num_instances, _round_robin_fn), 
    False, 
    buffer_size
)
```

where `_round_robin_fn` is:
```python
def _round_robin_fn(num_instances: int, idx_data) -> int:
        idx, _ = idx_data
        return idx % num_instances
```


Is that because you want to override `_DemultiplexerIterDataPipe.get_length_by_instance` ?

Reply:
Yeah. `demux` doesn't provide a valid length but this `DataPipe` should have a deterministic result. Let me know if you have better way to do so.

Reply:
noob question : what's "deterministic result" here mean? 

Especially I was somehow under the impression that in general try to avoid query `len(datapipe)` for large datapipe , since it probably needs to read all data and it's expensive?

Reply:
If the prior DataPipe provides a length, we should be able to get the length ahead of time.
For `demux` which uses a `classify_fn`, there is no way for us to know the length and the length can vary across epoches.

> in general try to avoid query `len(datapipe)` for large datapipe , since it probably needs to read all data and it's expensive?

It's true. But, we do provide `DataPipe` that can manually assign length to the DataPipe https://github.com/pytorch/data/blob/67afe0c36f5f93d34fde7026e7e95835c1160bb3/torchdata/datapipes/iter/util/header.py#L62